protocols/relay/src/behaviour/handler.rs

```diff
@@ -396,149 +408,173 @@ impl Handler {
+            let proto::HopMessage {
+                type_pb,
+                peer,
+                reservation: _,
+                limit: _,
+                status: _,
+            } = substream
+                .next()
+                .await
+                .ok_or(inbound_hop::FatalUpgradeError::StreamClosed)??;
+
+            let event = match type_pb {
+                Type::RESERVE => {
+                    let req = ReservationReq {
+                        substream,
+                        reservation_duration,
+                        max_circuit_duration,
+                        max_circuit_bytes,
+                    };
+
+                    Event::ReservationReqReceived {
+                        inbound_reservation_req: req,
+                        endpoint,
+                        renewed,
+                    }
+                }
+                Type::CONNECT => {
+                    let dst = PeerId::from_bytes(
+                        &peer.ok_or(inbound_hop::FatalUpgradeError::MissingPeer)?.id,
+                    )
+                    .map_err(|_| inbound_hop::FatalUpgradeError::ParsePeerId)?;
+                    let req = CircuitReq { dst, substream };
+
+                    Event::CircuitReqReceived {
+                        inbound_circuit_req: req,
+                        endpoint,
+                    }
+                }
+                Type::STATUS => return Err(inbound_hop::FatalUpgradeError::UnexpectedTypeStatus),
+            };
+
+            Ok(event)
         }
+        .boxed()
     }
 
-    fn on_fully_negotiated_outbound(
-        &mut self,
-        FullyNegotiatedOutbound {
-            protocol: (dst_stream, dst_pending_data),
-            info: outbound_open_info,
-        }: FullyNegotiatedOutbound<
-            <Self as ConnectionHandler>::OutboundProtocol,
-            <Self as ConnectionHandler>::OutboundOpenInfo,
-        >,
-    ) {
-        let OutboundOpenInfo {
-            circuit_id,
-            inbound_circuit_req,
-            src_peer_id,
-            src_connection_id,
-        } = outbound_open_info;
+    fn send_stop_message_and_process_result(
+        &self,
+        io: Stream,
+        stop_command: StopCommand,
+    ) -> BoxFuture<'static, Result<Event, FatalUpgradeError>> {
+        let msg = proto::StopMessage {
+            type_pb: proto::StopMessageType::CONNECT,
+            peer: Some(proto::Peer {
+                id: stop_command.src_peer_id.to_bytes(),
+                addrs: vec![],
+            }),
+            limit: Some(proto::Limit {
+                duration: Some(
+                    stop_command
+                        .max_circuit_duration
+                        .as_secs()
+                        .try_into()
+                        .expect("`max_circuit_duration` not to exceed `u32::MAX`."),
+                ),
+                data: Some(stop_command.max_circuit_bytes),
+            }),
+            status: None,
+        };
```

Comment:
Move this into the above described function to further reduce the noise in here.